src/Adapter/FirehoseAdapter.php

```diff
@@ -0,0 +1,163 @@
+<?php
+
+/**
+ * This file is part of graze/queue.
+ *
+ * Copyright (c) 2015 Nature Delivered Ltd. <https://www.graze.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ *
+ * @license https://github.com/graze/queue/blob/master/LICENSE MIT
+ *
+ * @link    https://github.com/graze/queue
+ */
+
+namespace Graze\Queue\Adapter;
+
+use Aws\Firehose\FirehoseClient;
+use Graze\Queue\Adapter\Exception\MethodNotSupportedException;
+use Graze\Queue\Adapter\Exception\FailedEnqueueException;
+use Graze\Queue\Message\MessageFactoryInterface;
+use Graze\Queue\Message\MessageInterface;
+
+/**
+ * Amazon AWS Kinesis Firehose Adapter.
+ *
+ * This method only supports the enqueue method to send messages to a Kinesiss
+ * Firehose stream
+ *
+ */
+final class FirehoseAdapter implements AdapterInterface
+{
+    const BATCHSIZE_SEND    = 100;
+
+    /** @var FirehoseClient */
+    protected $client;
+
+    /** @var array */
+    protected $options;
+
+    /** @var string */
+    protected $deliveryStreamName;
+
+    /**
+     * @param FirehoseClient $client
+     * @param string    $deliveryStreamName
+     * @param array     $options
+     */
+    public function __construct(FirehoseClient $client, $deliveryStreamName, array $options = [])
+    {
+        $this->client = $client;
+        $this->deliveryStreamName = $deliveryStreamName;
+        $this->options = $options;
+    }
+
+    /**
+     * @param MessageInterface[] $messages
+     *
+     * @throws MethodNotSupportedException
+     */
+    public function acknowledge(array $messages)
+    {
+        throw new MethodNotSupportedException(
+            'acknowledge',
+            $this,
+            $messages
+        );
+    }
+
+    /**
+     * @param MessageFactoryInterface $factory
+     * @param int                     $limit
+     *
+     * @throws MethodNotSupportedException
+     */
+    public function dequeue(MessageFactoryInterface $factory, $limit)
+    {
+        throw new MethodNotSupportedException(
+            'dequeue',
+            $this,
+            []
+        );
+    }
+
+    /**
+     * @param MessageInterface[] $messages
+     *
+     * @throws FailedEnqueueException
+     */
+    public function enqueue(array $messages)
+    {
+        $failed = [];
+        $batches = array_chunk($this->createEnqueueEntries($messages), self::BATCHSIZE_SEND);
+
+        foreach ($batches as $batch) {
+            $requestRecords = array_map(function ($a) {
+                return [
+                    'Data' => json_encode($a)
+                ];
+            }, $batch);
+
+            $request = [
+                'DeliveryStreamName' => $this->deliveryStreamName,
+                'Records'  => $requestRecords,
+            ];
+
+            $results = $this->client->putRecordBatch($request);
```

Comment:
Is there a top level failed result or does it always return a `RequestResponses`?

Reply:
http://docs.aws.amazon.com/aws-sdk-php/v3/api/api-firehose-2015-08-04.html#putrecordbatch

There's always a `RequestResponses` key in the result object. There is also a `FailedPutCount` element at the top-level in the response, but to know which messages have failed we'll need to inspect the `RequestResponses` object anyway.

Reply:
I am just thinking about what happens if the service is down, I assume guzzle would throw a `ServerException` on this call. As long as we are happy passing that back to the user, should be fine. Or we could `try{}catch{}` and wrap it in a `FailedEnqueueException`.

So the user of this can then just handle failed enqueue exceptions and re-queue happily.